### ISSUE TITLE
How to rescue `SocketError` when using RestClient?

### ISSUE THREAD

**Issue by ankitrg** [rest-client/rest-client#551]
I am facing a problem rescuing SocketError I have detailed the issue here http://stackoverflow.com/questions/40779163/how-to-rescue-socket-error-in-ruby/40779800#40779800

Can someone help me with this?

----

> I am using `RestClient` to make a network call in the ruby class. I am getting a SocketError whenever I am not connected to the internet. I have added a rescue block to catch the exception still I am not able to do so.
>
> the error message is: `SocketError (Failed to open TCP connection to api.something.com:443 (getaddrinfo: Name or service not known))`
> 
```
module MyProject
  class Client
    def get_object(url, params={})
      response = RestClient.get(url, {params: params})
    rescue SocketError => e
      puts "In Socket errror"
    rescue => e
      puts (e.class.inspect)
    end
  end
end
```
> The broad rescue gets called and print `SocketError`, but why the previous `rescue SocketError` is not triggered!
>
> Do you see something that I am missing?

**Comment by ab** [comment]
I'm not able to reproduce your issue. When I run your code, I see `In Scoket errror` printed.

What version of rest-client are you running?

**Comment by ankitrg** [comment]
Sorry for the typo :sob: 
and I am using `'rest-client', '~> 2.0'`, I believe this is the latest version.

**Comment by ab** [comment]
Hmm, do you have full sample code that I could run to demonstrate the issue?

**Comment by ankitrg** [comment]
I'm working on a project, give me some time, I'll ask someone and let you know if I can share the code with you. I'll update to ASAP.

**Comment by ab** [comment]
Ideally this wouldn't include project code, just a minimal amount of code needed to reproduce.

For example, this code:

```ruby
#!/usr/bin/env ruby
require 'rest-client'
require 'socket'

module MyProject
  class Client
    def get_object(url, params={})
      response = RestClient.get(url, {params: params})
    rescue SocketError => e
      puts "In Socket errror"
    rescue => e
      puts (e.class.inspect)
    end
  end
end

MyProject::Client.new.get_object('http://nonexistent.nodomain')
```

Results in this output:
```
In Socket errror
```

**Comment by ankitrg** [comment]
I thought it'll be difficult on you part as it involves facebook auth. Let me know if you need more clarity.

```
module FbAds
  class Client
    attr_reader :access_token

    def initialize(access_token)
      @access_token = access_token
    end

    def get_object(endpoint, params={})
      params.merge!({access_token: @access_token, format: 'json'})
      url = FbAds::Utils.build_url(endpoint)
      response = RestClient.get(url, {params: params})
      return JSON.parse(response.body)
    rescue RestClient::Exception => e
      if response_has_error?(e.response)
        handle_fb_api_errors(e.response)
      else
        raise FbError::RestClientException(e.response, e.message)
      end
    rescue SocketError => e
      puts "In Socket errror" # doesn't go here
    rescue => e
      puts (e.class.inspect) # still puts SocketError here
    end
  end
end
```

**Comment by AOphagen** [comment]
The same happens to me, not when my internet connection is disabled, but for a typo in the url - on purpose during testing. I mistyped `some.url.de` as `some.url.der` during testing - any non-existing url in a nested resource will bring up a SocketError for me, a non-existing page for an existing url will give a normal 404, that I can catch with rest-client.

To catch the resulting exception, I tried to go up thru your exceptions inheritances to the top (from `RestClient::RequestFailed` to `RuntimeError` class), not one of them grips `SocketError`. Maybe that helps a bit finding the reason?
